### PR TITLE
[Blockstore] Access to disposed object app crash on shutdown #404 

### PR DESCRIPTION
**Pull request by fassadlr**

### CODE REVIEW

File: Stratis.Bitcoin.Features.Consensus/ConsensusFeature.cs [hunk 1]
```diff
@@ -162,9 +176,7 @@ private void RunLoop()
                     {
                         this.chainState.HighestValidatedPoW = this.consensusLoop.Tip;
                         if (this.chain.Tip.HashBlock == block.ChainedBlock?.HashBlock)
-                        {
-                            this.consensusLoop.FlushAsync();
-                        }
+                            await this.consensusLoop.FlushAsync();
```
Comment (bokobza) [inline Stratis.Bitcoin.Features.Consensus/ConsensusFeature.cs]: So actually @dangershony and I are thinking that it's a bad idea for this to be async.
Can you please wait for the result of this call and check if it works?

File: Stratis.Bitcoin.Features.Consensus/ConsensusFeature.cs [hunk 2]
```diff
@@ -15,11 +15,18 @@
...
 
 namespace Stratis.Bitcoin.Features.Consensus
 {
     public class ConsensusFeature : FullNodeFeature
     {
+        /// <summary>Factory for creating and also possibly starting application defined tasks inside async loop.</summary>
+        private readonly IAsyncLoopFactory asyncLoopFactory;
+
+        /// <summary>The async loop we need to wait upon before we can shut down this feature.</summary>
+        private IAsyncLoop asyncLoop;
```
[the remark this answers is not in the document]
Comment (fassadlr) [reply Stratis.Bitcoin.Features.Consensus/ConsensusFeature.cs]: Not this, as we only set this field later, "readonly" fields can only be set in constructors ;)

File: Stratis.Bitcoin.Features.Consensus/ConsensusFeature.cs [hunk 3]
```diff
@@ -100,10 +109,10 @@ public override void Start()
 
             this.stakeChain?.Load().GetAwaiter().GetResult();
 
-            new Thread(this.RunLoop)
+            this.asyncLoop = this.asyncLoopFactory.Run($"Consensus Loop", async token =>
             {
-                Name = "Consensus Loop"
-            }.Start();
+                await RunLoop(this.nodeLifetime.ApplicationStopping);
```
[the remark this answers is not in the document]
Comment (fassadlr) [reply Stratis.Bitcoin.Features.Consensus/ConsensusFeature.cs]: Fixed

File: Stratis.Bitcoin.Features.Consensus/ConsensusFeature.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (bokobza) [reply Stratis.Bitcoin.Features.Consensus/ConsensusFeature.cs]: ah yes indeed, with the folding in git diff, I thought it was in the constructor :-)

File: Stratis.Bitcoin.Features.Consensus/ConsensusFeature.cs (hunk 1 above)
Comment (fassadlr) [reply Stratis.Bitcoin.Features.Consensus/ConsensusFeature.cs]: OK will fix 👍

**Comment by bokobza** [comment]
I think it looks good now.

**Pull request merged**